File: cpp-std/demo17a-reentrant-lock.cpp
/*
REENTRANT LOCKS (RECURSIVE MUTEXES)
Version A: Introduction to reentrant locks
*/


#include <iostream>
#include <thread>
#include <mutex>
using namespace std;



std::mutex mut;



void doTask() {
    mut.lock();
    cout << "First time acquiring the resource" << endl;

    mut.lock();
    cout << "Second time acquiring the resource" << endl;

    mut.unlock();
    mut.unlock();
}



int main() {
    auto th = std::thread(doTask);

    /*
    The thread th shall meet deadlock.
    So, you will never get output "Second time the acquiring resource".
    */

    th.join();

    return 0;
}

File: csharp/demo/demo17a-reentrant-lock.cs
/*
 * REENTRANT LOCKS (RECURSIVE MUTEXES)
 *
 * In C#, "lock" keyword and System.Threading.Monitor class allow re-entrancy.
 * That means they are reentrant locks by default.
 *
 * Version A: A simple example
 */
using System;
using System.Threading;



class Demo17A : IRunnable
{
    public void run()
    {
        const string resource = "resource";

        new Thread(() =>
        {
            lock (resource)
            {
                Console.WriteLine("First time acquiring the resource");

                lock (resource)
                {
                    Console.WriteLine("Second time acquiring the resource");
                }
            }
        }).Start();
    }
}
